text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1688 @@
+
+#### Constructability and Transmutation
+Transmutation supplies a mechanism for constructing instances of a type *without* invoking its implicit constructor, nor any constructors defined by the type's author. In the previous examples, it would be *unsafe* to transmute `[usize; 2]` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would violate memory safety:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+// [usize; 2] ⟶ NonEmptySlice
+let _: NonEmptySlice<'static, u8> = transmute!([0usize; 2]); // Compile Error: `NonEmptySlice<_, _>` is not constructible here.
+```
+If a field is private, then instantiating or modifying it via transmutation is not, generally speaking, safe.
+
+For transmutations where the destination type involves mutate-able references, the constructability of the *source* type is also relevant. Consider:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let arr = [0u8, 1u8, 2u8];
+let mut x = NonEmptySlice::from_array(&arr, 2);
+{
+    // &mut NonEmptySlice ⟶ &mut [usize; 2]
+    let y : &mut u128 = transmute!(&mut x) // Compile Error! `&mut NonEmptySlice` is not constructible here.
+    *y[0] = 0;
+    *y[1] = 0;
+}
+
+let z : NonEmptySlice<u8> = x;
+```
+If this example did not produce a compile error, the value of `z` would not be a safe instance of its type, `NonEmptySlice`, because `z.first()` would dereference a null pointer.
+
+#### Constructability and Scope
+Whether a type is fully implicitly constructable will depends on the *scope* in which that question is asked. Consider:
+```rust
+pub mod a {
+
+    #[repr(C)] pub struct Foo(private::Bar);
+
+    mod private {
+        #[repr(C)] pub struct Bar;
+    }
+
+    // `Foo` is fully implicitly constructible in this module.
+    const _: Foo = Foo { private::Bar };
+
+    // Thus, `Foo` is transmutable in this module!
+    const _: Foo = transmute!(());
+}
+
+pub mod b {
+    use super::a;
+
+    // `Foo` is NOT fully implicitly constructible in this module.
+    const _: Foo = a::Foo { a::private::Bar }; // Compile Error: the module `a::private` is private.
+
+    // Thus, `Foo` is NOT transmutable in this module:
+    const _: Foo = transmute!(()); // Compile Error: `Foo` is not constructible here.
+}
+```
+
+The `Constructible` marker trait makes the relationship between scope and constructability explicit:
+```rust
+pub unsafe trait Constructible<Scope>
+{}
+```
+A type `T` is `Constructible<Scope>`, if it is fully implicitly constructible in the scope represented by the type `Scope`. These `Scope` types are generated by the `Here!()` macro, which produces a type that uniquely identifies its invocation scope.
```

Comment:
I *think* `Constructible` is might just be a useful implementation and pedagogic device—not something users *actually* have any reason to interact with. I might mostly delete this paragraph. `Here!()` will still need some explanation, though.

I'll be curious if anybody can think of a use for `Constructible`. Types that are `Constructible` are instantiatable in const contexts, which is a somewhat neat correspondence.